### DISCUSSION TITLE
Error setting broker_read_url with Redis

### DISCUSSION THREAD

**Discussion by theelderbeever** [celery/celery#7464] [Q&A] [answered]
I am using Redis for my broker and result backend and wanted to leverage my read only instances in my Redis cluster. When I set the `broker_read_url` to my read only instance my workers throws exceptions about not being able to write on a read only instance

These are the configurations I have tried for my `Celery` app. 

```python
REDIS_PASSWORD = "password"
REDIS_HOST = "redis-master" # write instance
REDIS_READ_HOST = "redis-replicas" # read instances
REDIS_PORT = 6379

# This works
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"


# This works
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_write_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"


# This doesn't
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_write_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_read_url = f"redis://:{REDIS_PASSWORD}@{REDIS_READ_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"
```

This is the error thrown on worker start up
> redis.exceptions.ReadOnlyError: You can't write against a read only replica.

Any help that could be provided would be great!

## Environment:
AWS EKS
m5.2xlarge EC2 instances
Linux2 AMI
Python 3.10.2
Celery 5.2.3

**Comment by dan-licht** [comment]
@theelderbeever did you ever find a resolution to this? Facing the same issue

**Reply by theelderbeever** [reply]
> @theelderbeever did you ever find a resolution to this? Facing the same issue

@dan-licht not that I remember... And I haven't tried since. Sorry and best of luck.

**Reply by dan-licht** [reply]
@auvipy @Nusnus any suggestions on this? I feel like I'm most likely missing something simple here..

**Reply by Nusnus** [reply]
@dan-licht which version are you using?

**Reply by dan-licht** [reply]
using `celery[redis]==5.4.0` with gevent

**Reply by Nusnus** [reply, marked as answer]
This appears to be an inherited design limitation. Celery allows you to configure separate URLs for writing and reading (using `broker_write_url` for sending tasks and `broker_read_url` for receiving them). However, the worker uses the same connection it established for reading to also perform write operations.

For example, when a worker receives a task through the connection specified by `broker_read_url`, it must later acknowledge that task. Acknowledgment is a write operation. Since Celery uses the same connection for both fetching and acknowledging messages, if that connection points to a read‑only replica, the acknowledgment fails with a `ReadOnlyError`.

Consider the following configurations from the original issue:

```python
REDIS_PASSWORD = "password"
REDIS_HOST = "redis-master"  # write instance
REDIS_READ_HOST = "redis-replicas"  # read instance
REDIS_PORT = 6379

# This works
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"

# This works
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_write_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"

# This doesn't work
class Config:
    broker_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_write_url = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
    broker_read_url = f"redis://:{REDIS_PASSWORD}@{REDIS_READ_HOST}:{REDIS_PORT}/0"
    result_backend = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/1"
```

In the failing configuration, tasks are published to the writable master via `broker_write_url`, but the worker consumes tasks using `broker_read_url` (which points to a read‑only replica). Since the worker uses the same connection for receiving tasks and for sending acknowledgments, the write operation involved in acknowledging the task fails with a `ReadOnlyError`.

Thus, even though `broker_write_url` is used correctly for sending tasks, the fact that the worker consumes tasks via `broker_read_url`, which must also perform write operations, means that a read‑only replica is completely unsuitable for the worker’s broker connection.

**Reply by dan-licht** [reply]
Thanks for the thorough explanation! Just curious, do you foresee this limitation being addressed in an upcoming release?

**Reply by Nusnus** [reply]
It is not planned at all actually.
Contributions are always welcome though :)
